### ISSUE TITLE
type inference error on int/double literal in generics class method

### ISSUE THREAD

**Issue by Uchijo** [dart-lang/sdk#51780]
I think I found a issue with type inference on int/double literal in generics class method.

These are the environments I was able to reproduce the bug:
- Dart SDK Version: 2.18.6 (stable), 3.0.0-322.0.dev (master channel of dartpad.dev)
- OS: Windows
- Brower: Chrome, Firefox

this is a minimum code to reproduce the bug and the error message:

```dart
class Hoge<T> {
  Hoge(this.value);

  final T value;

  Hoge<T> self() => this;
}

class HogeBox<T> {
  HogeBox(this.hoge);
  final Hoge<T> hoge;
}

void fuga() {
  final box1 = HogeBox<double>(Hoge(1));
  final box2 = HogeBox<double>(Hoge(1).self());
  //                           ^ the argument type 'Hoge<int>' can't be assigned to the parameter type 'Hoge<double>'
}

```

The generic type T is double, as indicated by `HogeBox<double>`, so `Hoge(1)`'s generic type T should also be double. According to this, `Hoge(1).self()` in `HogeBox<double>` should also return `Hoge<double>`, as `Hoge(1)`'s generic type T is double. However, the error message indicates that `Hoge(1).self()` is of type `Hoge<int>`, not `Hoge<double>`.

This error affects, for example, `Tween#chain` in flutter.

**Comment by lrhn** [comment]
This is working as intended.

The top-down type context push-down does generally not go into receivers.
That means that `Hoge(1).self()` does not put `Hoge(1)` in a context expecting a `Hoge<double>`.

The `Hoge(1).self()` is in a context with context type `Hoge<double>`.
Then the type inference looks at `Hoge(1).self()`. To know what `.self()` means, it has to *first* find the type of `Hoge(1)`. At that point, it doesn't know that a `Hoge<double>` is needed, the receiver has no context type because we don't know what the relation between the outer context type and the `.self()` call is before we've figure out what type the `self` method has.
So `Hoge(1)`, with no context, creates a `Hoge<int>(1)`. Then calls `.self()` on that and finally complains because that's not assignable to `Hoge<double>`.

In short: Receivers of method invocations have no context types.

**Comment by Uchijo** [comment]
I now understand how type inference works in the Dart language and I am fully convinced.
Thank you for your detailed answer.
